### PR TITLE
Validate no unreferenced rows

### PR DESCRIPTION
**Pull request by tie-rack**
Pivotal story: [90230424](https://www.pivotaltracker.com/story/show/90230424)

Validate that submitted CSVs do not contain records that are not referenced by any other files. Each record that has an ID that is not referenced by another file produces a warning.

(Also, the "full good run" test data has been replaced with a more recent VA special election data set, slightly modified to produce no warnings. The old data set, it turns out, was overflowing with unreferenced records.)

### CODE REVIEW

File: src/vip/data_processor/validation/db/reverse_references.clj [hunk 1]
```diff
@@ -0,0 +1,61 @@
+(ns vip.data-processor.validation.db.reverse-references
+  (:require [vip.data-processor.validation.db.util :as util]
+            [korma.core :as korma]
+            [korma.sql.engine :as eng]))
+
+(defn find-referencing-column [table-name csv-spec]
+  (->> csv-spec
+       :columns
+       (filter #(= table-name (:references %)))
```
Comment (bwreid) [inline src/vip/data_processor/validation/db/reverse_references.clj]: I thought you didn't like this way of writing a function?  :)

File: src/vip/data_processor/validation/db/reverse_references.clj [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+(defn join-clause [from-table from-column to-table]
+  (let [from-column (util/column-name from-table from-column)
+        to-column (util/column-name to-table "id")]
+    (list '= from-column to-column)))
+
+(defn where-clause-part [table column]
+  (let [column (util/column-name table column)]
+    (list '= column nil)))
+
+(defn find-unreferenced-rows [tables table-id references-map]
```
Comment (bwreid) [inline src/vip/data_processor/validation/db/reverse_references.clj]: I think this function should at least have a doc string -- although more is better IMO.

File: src/vip/data_processor/validation/db/reverse_references.clj (hunk 1 above)
Comment (tie-rack) [inline src/vip/data_processor/validation/db/reverse_references.clj]: :sob: It's true.

File: src/vip/data_processor/validation/db/reverse_references.clj [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                 :let [table (:table csv-spec)
+                       references (find-referencing-tables table csv-specs)]
+                 :when (seq references)]
+             [table references])))
+
+(defn validate-no-unreferenced-rows-for-table [ctx [table-id references-map]]
+  (let [tables (:tables ctx)
+        unreferenced-rows (find-unreferenced-rows tables table-id references-map)]
+    (if (seq unreferenced-rows)
+      (assoc-in ctx [:warnings table-id "Unreferenced records"] unreferenced-rows)
```
Comment (bwreid) [inline src/vip/data_processor/validation/db/reverse_references.clj]: Is there a standard we've developed for these? My thought was they were all following: `[:higher-level-keyword "file-name" :error-type "additional details"]`

File: src/vip/data_processor/validation/db/reverse_references.clj (hunk 2 above)
Comment (tie-rack) [inline src/vip/data_processor/validation/db/reverse_references.clj]: I'll add some!

File: src/vip/data_processor/validation/db/reverse_references.clj (hunk 3 above)
Comment (tie-rack) [inline src/vip/data_processor/validation/db/reverse_references.clj]: You're right. We can find out from the `ctx` the file name the table was created for. I'll update that.

File: src/vip/data_processor/validation/db/reverse_references.clj (hunk 1 above)
Comment (tie-rack) [inline src/vip/data_processor/validation/db/reverse_references.clj]: Is `(comp (partial = table-name) :references)` more or less readable? ;)

Comment (bwreid) [inline src/vip/data_processor/validation/db/reverse_references.clj]: "readable."

Looks okay to me.  :)

**Pull request merged**